### DISCUSSION TITLE
Are lists of orientations (calcOrientations) representative of texture?

### DISCUSSION THREAD

**Discussion by ElliotCN** [mtex-toolbox/mtex#523] [Ask Anything] [answered]
Hi all,

I have a question about mtex and would be grateful for any help!

**Context:**
I have been using mtex to generate ideal textures for Crystal Plasticity Finite Element Model's (CPFEM's). I have been using the unimodal function with miller indices to produce a desired texture, in the form of an ODF, then using the calcOrientations to get a list of orientation that I can use in CPFEM's. 
 
**Issue:**
Once I have produced the list of orientations from the ODF, if I then use the list of orientations to create a new ODF it is not the same as the original ODF; it seems to give lower mrd values (see images below). If the list of orientations does not recreate the ODF it was made from, can the list of orientations be used as an accurate representation of the desired texture?

Thanks in advance,
Elliot

mtex (5.3) Script:

```matlab
%% parameters
crystalSym = crystalSymmetry('cubic');
specimenSym = specimenSymmetry('orthorhombic');
halfwidth = 10;
numOrientations = 1000;
modalOrientationHKL = [1,1,2];
modalOrientationUVW = [1,1,1];

 
%% Miller indices to ODF
psi = deLaValleePoussinKernel('halfwidth', halfwidth*degree);
ori = orientation.byMiller(modalOrientationHKL, modalOrientationUVW, crystalSym, specimenSym);
ODF = unimodalODF(ori, psi);
 
m = Miller({0,0,1},crystalSym);

figure();
plotPDF(ODF,m,'complete')

figure();
plot(ODF)


%% List of orientations to ODF
orientations = calcOrientations(ODF, numOrientations);
export(orientations, 'orientations.txt');

ODF_ori = loadODF_generic('orientations.txt', 'cs', crystalSym, 'ss', specimenSym,'ColumnNames',{'Euler 1' 'Euler 2' 'Euler 3'});

m = Miller({0,0,1},crystalSym);

figure();
plotPDF(ODF_ori,m,'complete')

figure();
plot(ODF_ori) 
```

![original vs orientaions ODF](https://user-images.githubusercontent.com/48714102/96005101-0af74e80-0e34-11eb-8556-3324b609a88a.jpg)

**Comment by ralfHielscher** [comment, marked as answer]
Hi Elliot,

when MTEX loads an ODF from individual orientations it implicitly performs a [density estimation](https://mtex-toolbox.github.io/DensityEstimation.html) to compute the ODF from the orientations. The crucial factor for this computation is the `halfwidth`. So if you do 
```matlab
ODF_ori = loadODF_generic('orientations.txt', 'cs', crystalSym, 'ss', specimenSym,...
  'ColumnNames',{'Euler 1' 'Euler 2' 'Euler 3'},'halfwidth',2*degree);
```
we will obtain a different ODF which might fit the original ODF better. Keep in mind that their is no one to one relationship between a set of individual orientations and an ODF.

Ralf.

**Reply by ElliotCN** [reply]
Hi Ralph,

Thank you for your reply. Adding the halfwidth of 2*degree does make it much closer to the original ODF.

If I specify the halfwidth for the ODF when generating the list of orientations why does the same halfwidth not reproduce the original ODF from the list of orientations?

Thanks,
Elliot

![halfwidth comparison](https://user-images.githubusercontent.com/48714102/96008881-ed2be880-0e37-11eb-9717-68f2b154d612.png)

**Comment by ralfHielscher** [comment]
Have you read the link I posted above? I tried to explain it there.

Ralf.

**Reply by ElliotCN** [reply]
Hi Ralph,

Apologies just realised that the density estimation was a link, I have read through it now. 

Just to check I have this correct would you agree with this?

As the half width at 10 degrees is larger than 2 degrees the density estimation is too smooth thus there is the blurring out effect I am seeing. When constructing an ODF from a list of orientations you need to take into consideration the effect of the halfwidth on the reconstruction.

Thanks,
Elliot

**Comment by kilir** [comment]
Hi,
if you load your orientations ori and estimate an odf using
```
psi = calcKernel(ori,'Method','KLCV')
odf = calcKernelODF(ori,'kernel',psi)
```
you should get a pretty close estimate of the original odf. At least for a reasonable number of orientations.

Cheers,
Rüdiger

**Comment by zmichels** [comment]
Hi Elliot,

Such an interesting issue in the context of modeling.

Does increasing the number of orientations (your numOrientations variable) yield any better results? What about if you use discreteSample() and the flag 'withoutReplacement’? 

example:
ono = discreteSample(odf,1000,'withoutReplacement’)
oyea = discreteSample(odf,1000)

Does ‘withoutReplacement’ give worse results, or better? Which sampling-replacement approach do you think makes more sense microstructurally when modeling a group of grains?

As to your latest question: I think this will almost always be the case for computing a density function from a discrete sample of a *different* density function – they are unlikely to ever be identical even with sampling of numerous points. Thats said, I would imagine that greatly increasing the number of orientation samples might help narrow the gap (and also whether or not you allow replacement, IDK), but perhaps that doesn’t work well with your model. And yet, I suppose the discrepancy may be useful context for designing your model and/or interpreting your model’s results. It seems to indicate that you would need many more grains to more-exactly approximate the original texture pattern. It would also be interesting to know or solve for which original-ODF psi will most often yield a 1000-grain sample that closest-matches the texture mrd you want to start with – because as you pointed out, the second ODF always has a lower mrd than the original texture.

Cheers,
Zach

**Comment by kilir** [comment]
I am not convinced that replacement/without replacement can be considered for random drawing from a continuous density function. a) you won't draw the exact same orientation twice anyways and b) if you think of a binned density fucntion, I guess you should draw more often the orientation from more densely populated bin.

The other thing is, that the halfwidth parameter of the unimodal component does not equal the halfwith parameter of the estimator. So even using a very large number of orientations but estimating but using the same parameter keeps the difference.

Cheers,
Rüdiger

**Comment by zmichels** [comment]
Thank you. That makes sense.

Cheers,
Zach
